mozetl/taar/taar_locale.py

```diff
@@ -0,0 +1,177 @@
+"""
+Bug 1396549 - TAAR Top addons per locale dictionary
+This notebook is adapted from a gist that computes the top N addons per
+locale after filtering for good candidates (e.g. no unsigned, no disabled,
+...) [1].
+
+[1] https://gist.github.com/mlopatka/46dddac9d063589275f06b0443fcc69d
+"""
+
+import boto3
+import click
+import json
+import logging
+import numpy as np
+
+from datetime import datetime
+from pyspark.sql import SparkSession
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+LOCALE_FILE_NAME = "top10_dict"
+
+
+def store_new_state(source_file_name, s3_dest_file_name, s3_prefix, bucket):
+    """Store the new json file containing current top addons per locale to S3.
+
+    :param source_file_name: The name of the local source file.
+    :param s3_dest_file_name: The name of the destination file on S3.
+    :param s3_prefix: The S3 prefix in the bucket.
+    :param bucket: The S3 bucket.
+    """
+    client = boto3.client('s3', 'us-west-2')
+    transfer = boto3.s3.transfer.S3Transfer(client)
+
+    # Update the state in the analysis bucket.
+    key_path = s3_prefix + s3_dest_file_name
+    transfer.upload_file(source_file_name, bucket, key_path)
+
+
+def get_addons(spark):
+    """ Longitudinal sample is selected and freshest ping chosen per client.
+    Only Firefox release clients are considered.
+    Columns are exploded (over addon keys)  to include locale of each addon
+    installation instance system addons, disabled addons, unsigned addons
+    are filtered out.
+    Sorting by addon-installations and grouped by locale.
+    """
+    return spark.sql("""
+        WITH sample AS (
+        SELECT client_id,
+        settings[0].locale AS locality,
+        EXPLODE(active_addons[0])
+        FROM longitudinal
+        WHERE normalized_channel='release'
+          AND build IS NOT NULL
+          AND build[0].application_name='Firefox'
+        ),
+
+        filtered_sample AS (
+        SELECT locality, key AS addon_key FROM sample
+        WHERE value['blocklisted'] = FALSE -- not blocklisted
+          AND value['type'] = 'extension' -- nice webextensions only
+          AND value['signed_state'] = 2 -- fully reviewed addons only
+          AND value['user_disabled'] = FALSE -- active addons only get counted
+          AND value['app_disabled'] = FALSE -- exclude compatibility disabled addons
+          AND value['is_system'] = FALSE -- exclude system addons
+          AND locality <> 'null'
+          AND key is not null
+        ),
+
+        country_addon_pairs AS (
+        SELECT
+        COUNT(*) AS pair_cnts, addon_key, locality
+        from filtered_sample
+        GROUP BY locality, addon_key
+        )
+
+        SELECT * FROM country_addon_pairs
+        ORDER BY locality, pair_cnts DESC
+    """)
+
+
+def compute_threshold(addon_df):
```

Comment:
You could have calculated this using [`spark.sql.DataFrame.approxQuantile`](http://spark.apache.org/docs/latest/api/python/pyspark.sql.html#pyspark.sql.DataFrame.approxQuantile), which takes a column, a list of quantile probabilities in [0, 1], and a precision term, and returns a list of approximated quantiles.

Reply:
Good point! I found that `np.percentile` behaves slightly differently because it tries to interpolate the data points (instead of returning the closest one) if the requested percentile doesn't exactly match a data point ([docs](https://docs.scipy.org/doc/numpy/reference/generated/numpy.percentile.html)). I had to tweak the test a little bit to account for that.